src/Umbraco.Web/Search/ExamineFinalComponent.cs

```diff
@@ -16,27 +18,65 @@ public sealed class ExamineFinalComponent : IComponent
         private readonly IExamineManager _examineManager;
         BackgroundIndexRebuilder _indexRebuilder;
         private readonly IMainDom _mainDom;
-        
-        public ExamineFinalComponent(IProfilingLogger logger, IExamineManager examineManager, BackgroundIndexRebuilder indexRebuilder, IMainDom mainDom)
+        private readonly ISyncBootStateAccessor _syncBootStateAccessor;
+        private readonly object _locker = new object();
+        private bool _initialized = false;
+
+        public ExamineFinalComponent(IProfilingLogger logger, IExamineManager examineManager, BackgroundIndexRebuilder indexRebuilder, IMainDom mainDom, ISyncBootStateAccessor syncBootStateAccessor)
         {
             _logger = logger;
             _examineManager = examineManager;
             _indexRebuilder = indexRebuilder;
             _mainDom = mainDom;
+            _syncBootStateAccessor = syncBootStateAccessor;
+
+            // must add the handler in the ctor because it will be too late in Initialize
+            // TODO: All of this boot synchronization for cold boot logic needs should be fixed in netcore
+            _syncBootStateAccessor.Booting += _syncBootStateAccessor_Booting;
         }
 
-        public void Initialize()
+        /// <summary>
+        /// Once the boot state is known we can see if we require rebuilds
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="e"></param>
+        private void _syncBootStateAccessor_Booting(object sender, SyncBootState e)
+        {
+            UmbracoModule.RouteAttempt += UmbracoModule_RouteAttempt;
+        }
+
+        private void UmbracoModule_RouteAttempt(object sender, RoutableAttemptEventArgs e)
         {
-            if (!_mainDom.IsMainDom) return;
+            UmbracoModule.RouteAttempt -= UmbracoModule_RouteAttempt;
 
-            _examineManager.ConfigureIndexes(_mainDom, _logger);
+            if (!_initialized)
+            {
+                lock (_locker)
```

Comment:
Does this lock need to be held past checking _initailized and IsMainDom? Could set an _initializing and return from the lock earlier to reduce blocking requests.

Reply:
This will only check on one request because the event is unbound as soon as it hits. The _initialize flag is now set as soon as it enters the double check, there won't be any blocking apart from the one request which just queues an operation.